Merge legacy `fetchGit`-builtin with the generic `fetchTree`-function

Ma27 opened the pull request:
The original idea was to implement a git-fetcher in Nix's core that
supports content hashes[1]. In #3549[2] it has been suggested to
actually use `fetchTree` for this since it's a fairly generic wrapper
over the new fetcher-API[3] and already supports content-hashes.

This patch implements a new git-fetcher based on `fetchTree` by
incorporating the following changes:

* Removed the original `fetchGit`-implementation and replaced it with an
  alias on the `fetchTree` implementation.

* Ensured that the `git`-fetcher from `libfetchers` always computes a
  content-hash and returns an "empty" revision on dirty trees (the
  latter one is needed to retain backwards-compatibility).

* The hash-mismatch error in the fetcher-API exits with code 102 as it
  usually happens whenever a hash-mismatch is detected by Nix.

* Removed the `flakes`-feature-flag: I didn't see a reason why this API
  is so tightly coupled to the flakes-API and at least `fetchGit` should
  remain usable without any feature-flags.

* It's only possible to specify a `narHash` for a `git`-tree if either a
  `ref` or a `rev` is given[4].

* It's now possible to specify an URL without a protocol. If it's missing,
  `file://` is automatically added as it was the case in the original
  `fetchGit`-implementation.

[1] https://github.com/NixOS/nix/pull/3216
[2] https://github.com/NixOS/nix/pull/3549#issuecomment-625194383
[3] https://github.com/NixOS/nix/pull/3459
[4] https://github.com/NixOS/nix/pull/3216#issuecomment-553956703

----

cc @edolstra @globin @xbreak @FRidh

Ma27 commented:
@domenkozar would you mind taking a look at this? :)

edolstra commented:
The reason for not adding a `narHash` flag to `fetchGit` is that we now have a more general builtin function `fetchTree`, which should already support `narHash`. So `fetchGit` / `fetchMercurial` are kind of legacy function that we keep for backward compatibility.

edolstra commented on src/libexpr/primops/fetchGit.cc:
```diff
@@ -34,12 +36,18 @@ static void prim_fetchGit(EvalState & state, const Pos & pos, Value * * args, Va
             throw EvalError(format("'url' argument required, at %1%") % pos);
+        if (narHash && !rev)
+            throw EvalError(
+                format("'narHash' can only be used for 'fetchGit' if 'rev' is set, at %1%") % pos
```
I think this is too strict. If you use a tag ref, it's fine to specify a `narHash` but not a `rev`.

Ma27 replied:
Agreed. Just realized that I allowed both revs and refs in the original PR.

Ma27 commented:
> The reason for not adding a narHash flag to fetchGit is that we now have a more general builtin function fetchTree, which should already support narHash. So fetchGit / fetchMercurial are kind of legacy function that we keep for backward compatibility.

I see. When I filed this patch, I wasn't' sure whether it's intended that `fetchGit` exists on its own. Should I modify this to use `fetchTree` entirely?

This would still require some modifications in the `git`-fetcher and the wrapper in `fetchTree` that passes arguments to the fetcher API.

edolstra commented:
Yes, please. All the other changes (like returning 102) look good to me.

Ma27 commented:
Rebased onto latest master and merged `fetchTree` and `fetchGit` into a single C++-function using the fetcher-API. More details can be found in the new commit-message.

@globin since you were quite interested in this and I had to change a lot, you may want to take another look at this :)

Ma27 commented:
ping @domenkozar @edolstra

domenkozar commented:
I'll leave it up to Eelco for final review.

edolstra commented:
Would you mind rebasing this on the flakes branch? I made some changes to the fetcher API there. (In particular `Input` is now basically just a set of attributes.)

Ma27 commented:
Yes, already on it :)

Ma27 commented:
@edolstra rebased onto `flakes` :)

edolstra commented on src/libexpr/primops/fetchTree.cc:
```diff
@@ -64,8 +97,15 @@ static void prim_fetchTree(EvalState & state, const Pos & pos, Value * * args, V
             state.forceValue(*attr.value);
-            if (attr.value->type == tString)
-                attrs.emplace(attr.name, attr.value->string.s);
+            if (attr.value->type == tPath)
```
=> `if (attr.value->type == tPath || attr.value->type == tString)` since `coerceToString` handles strings as well.

edolstra commented on src/libexpr/primops/fetchTree.cc:
```diff
@@ -48,10 +51,40 @@ void emitTreeAttrs(
+
+void validateFetcherAttrs(fetchers::Attrs attrs, const Pos &pos)
 {
-    settings.requireExperimentalFeature("flakes");
```
We need to keep this in `fetchTree`.

Ma27 replied:
Why though? At least `fetchGit` works (until now) without flakes.

edolstra replied:
Because `fetchTree` (not `fetchGit`) needs to be behind an experimental feature flag.

Ma27 replied:
Oh, you're absolutely right, fixed. Sorry.

edolstra commented on src/libexpr/primops/fetchTree.cc:
```diff
@@ -48,10 +51,40 @@ void emitTreeAttrs(
-static void prim_fetchTree(EvalState & state, const Pos & pos, Value * * args, Value & v)
+std::string fixURI(std::string uri, EvalState &state)
+{
+    state.checkURI(uri);
```
We can remove `checkURI` since the check can easily be by-passed.

Ma27 replied:
Without that most of the tests fail like this:

```
    + nix eval --raw --expr 'builtins.fetchTarball https://github.com/NixOS/patchelf/archive/master.tar.gz' --impure --restrict-eval
    error: --- RestrictedPathError --- nix
    access to URI 'https://github.com/NixOS/patchelf/archive/master.tar.gz' is forbidden in restricted mode
```

edolstra commented on src/libexpr/primops/fetchTree.cc:
```diff
@@ -36,6 +36,9 @@ void emitTreeAttrs(
     }
 
+    if (input.getType() == "git")
+        mkBool(*state.allocAttr(v, state.symbols.create("submodules")), maybeGetBoolAttr(input.attrs, "submodules").value_or(false));
```
This should only be done for `fetchGit`.

Ma27 replied:
Can you explain to me why this shouldn't happen for e.g. `fetchTree { type = "git"; }`?

edolstra replied:
We're trying to avoid returning unnecessary information from `fetchTree` since that might influence the evaluation result. But it probably doesn't matter much here.

edolstra commented on src/libexpr/primops/fetchTree.cc:
```diff
@@ -75,12 +115,24 @@ static void prim_fetchTree(EvalState & state, const Pos & pos, Value * * args, V
+    } else {
+        auto url = fixURI(state.coerceToString(pos, *args[0], context, false, false), state);
+
+        if (type && type.value() == "git") {
```
=> `type == "git"` should also work.

edolstra commented on src/libfetchers/git.cc:
```diff
@@ -269,6 +269,10 @@ struct GitInputScheme : InputScheme
 
+                // Backward compatibility: set 'rev' to
+                // 0000000000000000000000000000000000000000 for a dirty tree.
+                input.attrs.insert_or_assign("rev", "0000000000000000000000000000000000000000");
```
This should only be done for `fetchGit`.

Ma27 replied:
Do you want me to move this to the primop implementation? This is already the case for `git`-sources only.

edolstra replied:
What I mean is that it shouldn't be done for `fetchTree`, even for Git inputs.

Ma27 replied:
Fixed. However there's now no `rev` field if only a url for fetchGit is specified.

edolstra commented on src/libexpr/primops/fetchTree.cc:
```diff
@@ -48,10 +51,40 @@ void emitTreeAttrs(
+        auto ref = maybeGetStrAttr(attrs, "ref");
+        auto rev = maybeGetStrAttr(attrs, "rev");
+
+        if (narHash && !ref && !rev)
```
This validation should be done in `git.cc`, not `fetchTree.c`. Also, I think this test isn't really necessary, since people should be free to specify a narHash without a rev/ref. It could be fine in flake.nix, since the lock file will contain the rev...

Ma27 commented:
@edolstra rebased onto latest `flakes`-branch and fixed most of your comments (I added an answer/comment to the review-comments I didn't address (yet)).

Ericson2314 commented:
@edolstra Is there a rough ETA for merging the flakes branch? This change would be desirable for us too: if flakes will be merged soon that's fine, but if not, then maybe we could back-port something like the earlier `master`-targetted version of this one?

edolstra commented:
@Ericson2314 I can merge it after the RFC is merged.

Ma27 commented:
@edolstra may I ask why though? Those fetchers are even on non-flakes master atm.

Ma27 commented:
@edolstra rebased conflicts against master. Is there anything else tbd to get this merged?

Ma27 commented:
@edolstra the RFC is closed now - anything else we'd have to wait for until you can merge this? :)

Ma27 commented:
@edolstra resolved the merge conflict. Is there anything to discuss about the pending two comments you've left?

edolstra commented on tests/fetchGit.sh:
```diff
@@ -123,7 +129,7 @@ path2=$(nix eval --impure --raw --expr "(builtins.fetchGit file://$repo).outPath
 path3=$(nix eval --impure --raw --expr "(builtins.fetchGit $repo).outPath")
 # (check dirty-tree handling was used)
-[[ $(nix eval --impure --raw --expr "(builtins.fetchGit $repo).rev") = 0000000000000000000000000000000000000000 ]]
+[[ $(nix eval --impure --expr "(builtins.fetchGit $repo).rev or null") = null ]]
```
This breaks compatibility since there might be projects that expect that `rev` is always set. `fetchGit` (but not `fetchTree`) on dirty trees should continue to return `000...`.

Ma27 replied:
so, partly revert my change to ensure fetchGit returns an empty rev here?

edolstra replied:
Yes please.

Ma27 replied:
Fixed. However I'm well-aware that my current fix is fairly ugly since I didn't manage to allocate a new attr to `v` in `prim_fetchGit`. If you have a better solution, please tell me :)

edolstra commented:
@Ma27 Merged, thanks for all your work on this!

Ma27 commented:
You're welcome :) 

I'm surprised though that there's now a failingtest although I ran `make installcheck` several times locally. Tonight I have time to look into this.

Ma27 commented:
For the macos issue I asked in `#nixos-dev` whether someone with the needed hardware can take a look.
`nix-build -A checks.x86_64-linux` repeatedly builds localle.
